### PR TITLE
TextEditingDelta Mac

### PR DESCRIPTION
**Pull request by justinmc**
This PR will bring [TextEditingDeltas](https://docs.google.com/document/d/1PXNd_LwJudiUTnCM2AXRxbd98_p4BlFf6qFhEhv3wvs/edit#heading=h.pub7jnop54q0), originally introduced by @Renzo-Olivares, to Mac.  This is very similar to the iOS implementation (https://github.com/flutter/engine/pull/28418).

Partial fix for https://github.com/flutter/flutter/issues/87972

### CODE REVIEW

File: shell/platform/darwin/macos/framework/Source/FlutterTextEditingDelta.h
Comment (justinmc) [reply shell/platform/darwin/macos/framework/Source/FlutterTextEditingDelta.h]: The reusable class works, I've modified this PR to use it for both Mac and iOS.  I'll use it in subsequent PRs on other platforms as well.

**Comment by justinmc** [comment]
@LongCatIsLooong Ready for re-review.

File: shell/platform/common/text_editing_delta.h
Comment (LongCatIsLooong) [reply shell/platform/common/text_editing_delta.h:26]: Also can this be a struct since the deltas are short-lived and immutable for the most part? Not familiar with C++ not sure what the best practice here is.

File: shell/platform/darwin/macos/framework/Source/FlutterTextInputPlugin.mm
Comment (LongCatIsLooong) [inline shell/platform/darwin/macos/framework/Source/FlutterTextInputPlugin.mm:636]: According to the comment `string` here can be a `NSAttributedString` which is not a subtype of `NSString` and does not respond to the `UTF8String` selector I think, so it will probably crash the app if a `NSAttributedString` is passed to this method.

File: shell/platform/darwin/macos/framework/Source/FlutterTextInputPlugin.mm
Comment (LongCatIsLooong) [inline shell/platform/darwin/macos/framework/Source/FlutterTextInputPlugin.mm:451]: Is this method being called anywhere?

**Review by LongCatIsLooong — COMMENTED** [review]
I'm a little concerned about the (seemingly?) unnecessary string encoding and decoding otherwise LGTM. I don't know C++ so @cbracken would you mind taking a look?

File: shell/platform/darwin/macos/framework/Source/FlutterTextInputPlugin.mm
Comment (justinmc) [reply shell/platform/darwin/macos/framework/Source/FlutterTextInputPlugin.mm:451]: Looks like it's not on Mac, good catch.  Removing.

File: shell/platform/common/text_editing_delta.h
Comment (justinmc) [reply shell/platform/common/text_editing_delta.h:26]: @cbracken Can you advise on these two points?

 1. I've made the class into a struct for now.  I don't know if it makes a difference in this case though.  I'm interested to know if there are best practices to follow.

 2. About UTF8/16.  It seems like Mac uses text_input_model, which internally stores the text in UTF16, but converts to UTF8 when accessed in GetText ([code](https://github.com/flutter/engine/blob/387563aecb9ff7ace912fdf7bc297700f3913641/shell/platform/common/text_input_model.cc#L292-L296)).  iOS stores the text in NSMutableString ([code](https://github.com/flutter/engine/blob/387563aecb9ff7ace912fdf7bc297700f3913641/shell/platform/darwin/ios/framework/Source/FlutterTextInputPlugin.h#L60)), which is UTF16.  My TextEditingDelta class in this PR stores the text in a UTF8 string.  We'd like to minimize conversions, maybe by letting TextEditingDelta take any type by using a template?

Comment (iskakaushik) [reply shell/platform/common/text_editing_delta.h:26]: 1. For structs vs classes, take a look at https://google.github.io/styleguide/cppguide.html#Structs_vs._Classes. In C++ they are almost identical. In this case we could go with a struct, but I don't think using a class is particularly bad as well.
2. I'm not sure where this class gets instantiated, but depending on how much control we have we can either go with [u16string](https://www.cplusplus.com/reference/string/u16string/) or taking in a [basic_string](https://www.cplusplus.com/reference/string/basic_string/) which would cover both utf-8, 16.

Comment (cbracken) [reply shell/platform/common/text_editing_delta.h:26]: Apologies for missing this earlier. Agreed on point 1.

Point 2 will depend on how the string is encoded. If you're sending a UTF-8 string, use `std::string`. If the string is UTF-16, then use `std::u16string`. You'll notice in our Windows APIs that we make extensive use of `std::wstring` which is a string of `wchar_t`. In general, `std::wstring` and `std::u16string` are _NOT_ the same, but on Windows, they are.

In other places in the code that live across multiple platforms, you'll see that we provide multiple setters/getters. An example is [TextInputModel](https://github.com/flutter/engine/blob/main/shell/platform/common/text_input_model.h). In that case, we store the text as UTF-16 and provide setters for UTF-8 and UTF-16.

Comment (iskakaushik) [reply shell/platform/common/text_editing_delta.h:26]: +1 to what @cbracken said

File: shell/platform/darwin/macos/framework/Source/FlutterTextInputPlugin.mm
Comment (LongCatIsLooong) [inline shell/platform/darwin/macos/framework/Source/FlutterTextInputPlugin.mm:401]: Does this work? `c_str()` returns a UTF-8 string but the `NSDictionary` literal should only take `NSString` I think?

File: shell/platform/common/text_editing_delta.h
Comment (LongCatIsLooong) [reply shell/platform/common/text_editing_delta.h:26]: Thanks for the insights!

Regarding encoding, if you take a `NSString` and convert it to UTF-8 to store in a `TextEditingDelta`, when you send the `TextEditingDelta` to the framework the string needs to be convert back to `NSString` (I think?), and then `NSJSONSerialization` converts it to JSON (utf-8), unless `NSString` is smart enough to recognize the string is immutable and only does UTF-8 => UTF-16 conversion lazily.

Comment (cbracken) [reply shell/platform/common/text_editing_delta.h:26]: > unless NSString is smart enough to recognize the string is immutable and only does UTF-8 => UTF-16 conversion lazily. 

NSString has a fairly complex internal representation at least last I remember. For short ASCII-ish strings, it uses a [tagged pointer representation](https://mikeash.com/pyblog/friday-qa-2015-07-31-tagged-pointer-strings.html), for longer strings, it allocates a buffer and stores as a UTF-16 representation.

File: shell/platform/darwin/macos/framework/Source/FlutterTextInputPlugin.mm
Comment (LongCatIsLooong) [reply shell/platform/darwin/macos/framework/Source/FlutterTextInputPlugin.mm:401]: Oh the `@()` will be turned into `-[NSString stringWithUTF8String]` by the compiler.

File: shell/platform/common/text_editing_delta.h
Comment (justinmc) [reply shell/platform/common/text_editing_delta.h:26]: Is there any action to take here on the utf8/16 topic then?  I could make TextEditingDelta store the text as std::u16string, but I'm unclear if that would better, worse, or the same in terms of number of conversions including both iOS and Mac.

Comment (cbracken) [reply shell/platform/common/text_editing_delta.h:26]: Given that this is a private, embedder-internal representation, I'd lean towards using storing as `std::u16string` and supporting both `std::string` and `std::u16string` setters/constructor. 

On Windows, that will almost certainly result in less conversions and it may also be the case with NSString, which, from memory, uses a UTF-16 representation for longer strings where it isn't using a tagged pointer representation.

Comment (justinmc) [reply shell/platform/common/text_editing_delta.h:26]: That's right, we can change this in the future without breaking the framework.

Sounds good to me, I'll port this over to store the text in TextEditingDelta as std::u16string.  I'll do it in one commit so it's easy to review.  Thanks for the clarification.

File: shell/platform/common/text_editing_delta.h
Comment (justinmc) [inline shell/platform/common/text_editing_delta.h:7]: The use of `codecvt_utf8_utf16` caused a deprecation warning on Windows that wanted me to use `wstring`, but I've silenced it since this code needs to work on multiple platforms.

**Comment by justinmc** [comment]
Ready for re-review.  I've updated the TextEditingDelta struct to store the strings as utf-16 while providing setters and constructors for both utf 8 and 16.

**Comment by skia-gold** [comment]
Gold has detected about 1 new digest(s) on patchset 28.
View them at https://flutter-engine-gold.skia.org/cl/github/29036

File: shell/platform/common/text_editing_delta.h
Comment (cbracken) [reply shell/platform/common/text_editing_delta.h:7]: Yep; we enable this in a few places. That said, instead of defining it here, please define this in shell/platform/common/BUILD.gn as we do [here](https://github.com/flutter/engine/blob/bba38c0b83122ec9b174271e74499d739416208b/shell/platform/common/BUILD.gn#L54-L57)

File: shell/platform/common/text_editing_delta.h
Comment (cbracken) [inline shell/platform/common/text_editing_delta.h:16]: Add doc comments to this struct and the methods.

File: shell/platform/common/text_editing_delta.h
Comment (cbracken) [inline shell/platform/common/text_editing_delta.h:26]: Here and below, methods and functions should be named in `UpperCamelCase()`. 

Simple property getters/setters can be named `foo_bar()` and `set_foo_bar()`.

Ref: https://google.github.io/styleguide/cppguide.html#Function_Names

File: shell/platform/common/text_editing_delta.h
Comment (cbracken) [inline shell/platform/common/text_editing_delta.h:20]: It looks like `TextEditingDelta`s are copyable types. In this case, you should define:
* A copy constructor: `TextEditingDelta(const TextEditingDelta& other)`
* operator=: `TextEditingDelta& operator=(const TextEditingDelta& other)`

In this case, you should be able to declare them as `default`.

Ref: https://google.github.io/styleguide/cppguide.html#Copyable_Movable_Types

File: shell/platform/common/text_editing_delta.h
Comment (cbracken) [inline shell/platform/common/text_editing_delta.h:17]: This line is not needed since struct members are public by default.

File: shell/platform/common/text_editing_delta.h
Comment (cbracken) [inline shell/platform/common/text_editing_delta.h:18]: Here and below, parameters, fields, and variables should be formatted as `text_before_change`.

Ref: https://google.github.io/styleguide/cppguide.html#Variable_Names

File: shell/platform/common/text_editing_delta.cc
Comment (cbracken) [inline shell/platform/common/text_editing_delta.cc:16]: Here and below: I'd be tempted to get rid of `SetDeltas` altogether and just move initialisation of the instance variables to the initialiser list:

```cc
TextEditingDelta::TextEditingDelta(std::u16string textBeforeChange,
                                   TextRange range,
                                   std::u16string text) 
  : oldText_(textBeforeChange),
    deltaText_(text),
    deltaStart_(range.start()),
    deltaEnd_(range.start() + range.length()) {
} 
```
Note that the order in the initialiser list _must_ be the same as the order of declaration in the struct.

File: shell/platform/common/text_editing_delta.h
Comment (cbracken) [inline shell/platform/common/text_editing_delta.h:28]: Here and elsewhere: for parameters, use `const std::string& param_name`.

(Yes, we could use `std::string_view` and `std::u16string_view`, but I'd rather hold off and do a pass over the whole codebase where we do something consistent -- for now, similar classes like `TextInputModel` use string refs, so best for us to to keep consistent with them)

File: shell/platform/common/text_editing_delta.h
Comment (cbracken) [inline shell/platform/common/text_editing_delta.h:48]: If we do keep this method around, `s/txt/text/` on this, plus switch to `words_separated_by_underscores` form for all of these.

File: shell/platform/common/text_editing_delta.h
Comment (cbracken) [inline shell/platform/common/text_editing_delta.h:57]: Similarly here, use `const std::u16string& text`.

File: shell/platform/common/text_editing_delta.h
Comment (cbracken) [inline shell/platform/common/text_editing_delta.h:67]: Since these two functions don't reference any instance variables they should just be static functions in the C++; that said they're basically just two-liners, and I bet deleting these functions and just using `wstring_convert` inline in the two places it's used would actually be less lines of code 😉.

e.g. of where else we do this:
https://github.com/flutter/engine/blob/bba38c0b83122ec9b174271e74499d739416208b/shell/platform/common/text_input_model.cc#L36-L42

File: shell/platform/common/text_editing_delta_unittests.cc
Comment (cbracken) [inline shell/platform/common/text_editing_delta_unittests.cc:34]: Here and above: These could probably be `EXPECT_EQ` since these checks aren't interdependent on each other.

Ref: https://testing.googleblog.com/2008/07/tott-expect-vs-assert.html

File: shell/platform/common/text_editing_delta_unittests.cc
Comment (cbracken) [inline shell/platform/common/text_editing_delta_unittests.cc:15]: `flutter::` prefix isn't necessary here since you're already in the `flutter` namespace.

Just:

```
TextRange range(0, 4);
```

**Review by cbracken — CHANGES_REQUESTED** [review]
I've just reviewed the C++ bits -- haven't looked yet looked at the macOS specific Obj-C side.

File: shell/platform/common/text_editing_delta.h
Comment (justinmc) [reply shell/platform/common/text_editing_delta.h:17]: Let me know if I've missed any that usually should have doc comments.  I skipped some things that seemed obvious.

File: shell/platform/common/text_editing_delta.cc
Comment (justinmc) [reply shell/platform/common/text_editing_delta.cc]: I was able to get rid of SetDeltas, thanks.

File: shell/platform/common/text_editing_delta.h
Comment (justinmc) [reply shell/platform/common/text_editing_delta.h:85]: I ended up keeping these methods around because I needed to use them in the initializers, where I think it would be harder to directly use wstring_convert.  Let me know there is a slicker 1 liner way to use them that would be better.

File: shell/platform/common/text_editing_delta.h
Comment (justinmc) [reply shell/platform/common/text_editing_delta.h:7]: Isn't that definition you linked exactly where I'd need to define this for this file?  text_editing_delta.h and cc are included in that block in this PR.  Or am I misunderstanding?  I still seemed to get a failure in the Windows tests caused by this.

**Comment by justinmc** [comment]
@cbracken Ready for re-review.  Thanks for all of the suggestions and references as I figure out C++ at Google.

I think I took all of your suggestions except for two that I left questions on above.

**Comment by skia-gold** [comment]
Gold has detected about 253 new digest(s) on patchset 29.
View them at https://flutter-engine-gold.skia.org/cl/github/29036

File: shell/platform/common/text_editing_delta.h
Comment (cbracken) [reply shell/platform/common/text_editing_delta.h:7]: Ah yeah -- looks like that is indeed the exact target. Will try patching this in to see if I can repro.

File: shell/platform/common/text_editing_delta.h
Comment (cbracken) [reply shell/platform/common/text_editing_delta.h:85]: Nope -- no one-liner trick here unfortunately. SGTM!

File: shell/platform/common/text_editing_delta_unittests.cc
Comment (cbracken) [inline shell/platform/common/text_editing_delta_unittests.cc:14]: nit: `replacement_text`

File: shell/platform/common/text_editing_delta.cc
Comment (cbracken) [inline shell/platform/common/text_editing_delta.cc:11]: `const std::u16string& text_before_change`

File: shell/platform/common/text_editing_delta.h
Comment (cbracken) [inline shell/platform/common/text_editing_delta.h:30]: You could mark this default and kill off the one in the cc file.

File: shell/platform/common/text_editing_delta_unittests.cc
Comment (cbracken) [inline shell/platform/common/text_editing_delta_unittests.cc:4]: Nit: this should be a blank like (kill the `//`)

**Review by cbracken — COMMENTED** [review]
LGTM on the C++ side modulo a couple minor nits. Will take a look at the macOS implementation after lunch!

**Comment by justinmc** [comment]
Not sure why gold picked up 253 new gold images...  I've merged in master and hopefully that fixes it.

File: shell/platform/darwin/macos/framework/Source/FlutterTextInputPlugin.mm
Comment (cbracken) [inline shell/platform/darwin/macos/framework/Source/FlutterTextInputPlugin.mm:444]: What about `clampRange:(NSRange*)range forText:(NSString*)text` in case we ever use this on e.g. mark text?

**Review by cbracken — APPROVED** [review]
lgtm!

File: shell/platform/darwin/macos/framework/Source/FlutterTextInputPlugin.mm
Comment (justinmc) [reply shell/platform/darwin/macos/framework/Source/FlutterTextInputPlugin.mm:444]: Actually it looks like I've removed all usage of this 🤷.  I'll remove it altogether then.

**Comment by justinmc** [comment]
@cbracken I'm not authorized to merge this PR apparently, can you do it?

**Comment by cbracken** [comment]
Ah -- the issue is that you're trying to merge against `master`. We switched the default branch to `main` last night.

**Comment by justinmc** [comment]
Ah there we go, thanks!

**Pull request merged**
